### PR TITLE
Support passing filters to `discovery.NewHealthCheck(...)`

### PR DESCRIPTION
**Pull request by timvaillancourt**
## Description

This PR resolves https://github.com/vitessio/vitess/issues/16074 by allowing filters to be passed into `discovery.NewHealthCheck(...)` _(vs. being inherited from flags only)_ as `vtgate` and `txthrottler` _(in `vttablet`)_ have very different expectations of filtering healthchecks

This change makes "filtering" a 1st-class feature of the healthcheck logic and both use-cases can pass in their required filter(s). This ability to pass filters will be used to filter the healthcheck streams for the `txthrottler` to the local-shard only, in order to resolve https://github.com/vitessio/vitess/issues/16074

## Related Issue(s)

  - Fixes: https://github.com/vitessio/vitess/issues/16074, where `txthrottler` monitors an entire cell instead of the single-shard it should concern itself with

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Comment by mattlord** [comment]
You can address the failing test by merging in main now:
```
git fetch --all
git merge origin/main
```
 🙏

**Comment by timvaillancourt** [comment]
Correction: this is not a regression. [I submitted a fix for this ](https://github.com/vitessio/vitess/pull/13062#issuecomment-1789588796) earlier but it was not merged due to other refactoring 🤦 

Our v14 release doesn't hit this due to using `discovery.LegacyHealthCheck` (now gone)

File: go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go:59]: I'm curious how this work is cancelled and cleaned up w/o a cancelable context? It's not new here in this PR, I know.

File: go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go:349]: IMO it's better to remove the named return parameter and return here:
```suggestion
		return ts.initHealthCheckStream(topoServer, target)
	}
	return nil
```

Comment (mattlord) [reply go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go:349]: We should take this opportunity to replace: `if !reflect.DeepEqual(knownCells, ts.healthCheckCells) {` with:
```
if !slices.Equal(knownCells, ts.healthCheckCells) {
```

**Review by mattlord — COMMENTED** [review]
Looks good! I only had some minor comments/questions/requests. Let me know what you think. ❤️

File: go/vt/discovery/healthcheck.go
Comment (mattlord) [reply go/vt/discovery/healthcheck.go:307]: From looking at the binary flags, I can see that `--tablet_filters` can be supplied for `vtgate`, `vttablet`, and `vtctld`.  And `--keyspaces_to_watch can be supplied for `vtgate` and `vtctld`. And finally `--tablet-filter-tags` can be supplied for `vtgate` and `vtcombo`. So that answers this part.

File: go/vt/discovery/healthcheck_test.go
Comment (mattlord) [inline go/vt/discovery/healthcheck_test.go:1]: We should have a unit and/or endtoend test that confirms this new behavior — specifically that the txthrottler only initiates health streams from tablets within its shard. It should fail on main and pass on the branch. R

File: go/vt/discovery/healthcheck.go
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck.go:307]: @mattlord sure, I can add those function comments

This block is a mostly-unchanged copy/paste of existing logic _(from `NewHealthCheck` - lower in the file)_, so the flag and functionality are at least partially documented already. Any suggestion what to add?

File: go/vt/discovery/healthcheck.go
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck.go:110]: @mattlord this is a copy/paste of an existing error from `. NewHealthCheck()`. Any concern with changing an existing error?

File: go/vt/discovery/healthcheck.go
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck.go:316]: @mattlord this error is also copy/pasted from existing code

File: go/vt/discovery/healthcheck.go
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck.go:355]: @mattlord yes that is what it's doing. The code comment seems clear to me, any suggestion what to change?

File: go/vt/discovery/healthcheck_test.go
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck_test.go:1]: @mattlord good idea, I'll add this as a unit test somehow 👍

File: go/vt/discovery/healthcheck_test.go
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck_test.go:116]: @mattlord oh nice, TIL 👍

File: go/vt/discovery/healthcheck.go
Comment (mattlord) [reply go/vt/discovery/healthcheck.go:316]: It's fine to change it, but also not required.

File: go/vt/discovery/healthcheck.go
Comment (mattlord) [reply go/vt/discovery/healthcheck.go:110]: We should change it as it's wrong. This got missed when we changed everything from single to double dashes.

File: go/vt/discovery/healthcheck.go
Comment (mattlord) [reply go/vt/discovery/healthcheck.go:307]: I think you added the tablet filter flags? I'm not sure how it's supposed to work either as I've never seen any docs in the code or the website for it. 🙂 Maybe I missed it?

File: go/vt/discovery/healthcheck.go
Comment (mattlord) [reply go/vt/discovery/healthcheck.go:355]: Suggestion is in my comment.

File: go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go
Comment (timvaillancourt) [reply go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go:59]: @mattlord I don't have context (pun intended) on that. @ejortegau do you?

Comment (timvaillancourt) [reply go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go:59]: @mattlord it appears that context [is passed to the undelying topo watcher only](https://github.com/vitessio/vitess/blob/149bf92fe90a1bc1592d8cf444c567ffad083864/go/vt/discovery/healthcheck.go#L383). [Context on this change is here](https://github.com/vitessio/vitess/pull/10542)

File: go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go
Comment (timvaillancourt) [reply go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go]: @mattlord agreed on `slices`, this didn't exist in the runtime when we wrote this

This was added to a different refactoring-focused PR I already made here: https://github.com/vitessio/vitess/pull/16248

File: go/vt/discovery/healthcheck.go
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck.go:355]: @mattlord oh thanks, it's your suggestion that is the clear one. Updating

File: go/vt/discovery/healthcheck.go
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck.go:316]: @mattlord sounds good

For clarity I tend to separate bug fixes for a specific issue and code refactoring, but this is good to know. Updating the errors

**Review by harshit-gangal — COMMENTED** [review]
changes looks good. we should test the new changes.

File: go/vt/discovery/healthcheck.go
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck.go]: Updated

File: go/vt/discovery/healthcheck.go
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck.go]: Updated

File: go/vt/discovery/healthcheck.go
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck.go]: Updated

File: go/vt/discovery/healthcheck_test.go
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck_test.go]: Updated

File: go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go
Comment (timvaillancourt) [reply go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go]: Updated via https://github.com/vitessio/vitess/pull/16248

File: go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go
Comment (timvaillancourt) [reply go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go]: @mattlord reworked the way `context.Context` is passed [here](https://github.com/vitessio/vitess/pull/16170/commits/1f7744ee1540cdb6aebf604f6a2df4ec0914acba), possibly just for clarity. Now the state struct that lives between `.Open()` and `.Close()`-only holds a context/cancel func

**Comment by timvaillancourt** [comment]
@mattlord / @harshit-gangal I've added a basic test for tablet filtering to `TestCellTabletsWatcher`. Also each tablet filter implementation is tested in a new test from this PR: `TestNewVTGateHealthCheckFilters`

Let me know if we'd like to test anything further 🙇!

File: .github/CODEOWNERS
Comment (mattlord) [inline .github/CODEOWNERS:36]: I wanted to point out that being a code owner is a responsibility. Meaning that you will be selected as a reviewer on PRs that modify this code and you are expected to review those PRs in a timely fashion as an owner of that code/area. I'm noting this as I have not seen you very active as a reviewer for the other areas in which you are already listed as a code owner. I would recommend not adding yourself to more areas until you can be more involved with reviews (it takes a lot of time and effort).

File: go/vt/discovery/healthcheck_test.go
Comment (mattlord) [inline go/vt/discovery/healthcheck_test.go:87]: How does this confirm any specific behavior? i.e. that we filter out any tablets in that `ks1/-80` shard?

File: go/vt/discovery/healthcheck_test.go
Comment (mattlord) [inline go/vt/discovery/healthcheck_test.go:92]: Same thing here, and for the others. It's not obvious to me how we're testing the specific behavior (I do see that we test that we get the expected *type* of filter(s). But don't we want to test that 1) we get the specific filters we expect and 2) the filtering works as expected (meaning that specific tablets are ignored or filtered out?

File: go/vt/discovery/healthcheck_test.go
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck_test.go:87]: @mattlord this test confirms the right filter is provided. The individual filter implementations are tested here:
1. [`FilterByShard`](https://github.com/vitessio/vitess/blob/main/go/vt/discovery/topology_watcher_test.go#L301-L371)
2. [`FilterByKeyspace`](https://github.com/vitessio/vitess/blob/main/go/vt/discovery/topology_watcher_test.go#L392-L463)
3. [`FilterByTags`](https://github.com/vitessio/vitess/blob/main/go/vt/discovery/topology_watcher_test.go#L581-L606)

File: go/vt/discovery/healthcheck_test.go
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck_test.go:92]: @mattlord those tests are in `topology_watcher_test.go`

File: go/vt/discovery/topology_watcher_test.go
Comment (timvaillancourt) [reply go/vt/discovery/topology_watcher_test.go:308]: @mattlord good point, added this

File: go/vt/discovery/healthcheck_test.go
Comment (mattlord) [reply go/vt/discovery/healthcheck_test.go:87]: Thanks! I must be blind 🤣

**Review by mattlord — APPROVED** [review]
Thanks, @timvaillancourt !

**Comment by timvaillancourt** [comment]
@mattlord thanks for the review!

You mentioned documentation for the various filtering earlier. The command line flags are documented of course, but it seems the concept of filtering in general is not explained in the overall docs

Let me know if you were thinking about an actual doc section and if you had a suggestion where it should be added 🙇

**Review by deepthi — APPROVED** [review]

**Comment by deepthi** [comment]
@timvaillancourt every one of the commits in this PR is missing a sign-off. You will need to amend the commits and force-push. Once that is done, you should be able to push the button to merge. Please do "Squash and merge", that's our convention.

**Pull request merged**
